debian/manpages/electricsheep-voter.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH ELECTRICSHEEP-VOTER 1 "June 17, 2005"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
electricsheep-voter \- voter tool for electricsheep(1)
.SH SYNOPSIS
.B electricsheep-voter
.RI [ options ]
.SH DESCRIPTION
electricsheep-voter 
is a tool used by electricsheep when voting for a flame.
.SH OPTIONS
This program should be called with the following command line:
electricsheep-voter window nick url history_file dream_server leave_prefix uid
.br
.SH AUTHOR
This manual page was written by Alexis Sukrieh <sukria@sukria.net> 
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
electricsheep (2.6.8-5) unstable; urgency=low

  * Removed the debian revision from the version string so sheeps can be
    downloaded from the server.
    (closes: #391990)

 -- Alexis Sukrieh <sukria@debian.org>  Wed, 11 Oct 2006 07:54:36 +0200

electricsheep (2.6.8-4) unstable; urgency=high

  [ Jérémy Bobbio ]

  * debian/patches/04_Composite.dpatch
    + tells mpeg2dec_onroot to skip ARGB visuals
    when trying to find the best visual. Fix an issue when running inside a
    Composite-enabled X environment.
    (closes: #390902)

  [ Alexis Sukrieh ]

  * urgency set to high, for closing a RC bug in testing (BSP).

 -- Alexis Sukrieh <sukria@debian.org>  Sat,  7 Oct 2006 16:49:00 +0200

electricsheep (2.6.8-3) unstable; urgency=low

  * mpeg2dec/libmpeg2/idct_altivec.c
    Applied patch from p2@mind.be for fixing the FTBFS issue on ppc
    architectures.
    (closes: #386673)

 -- Alexis Sukrieh <sukria@debian.org>  Mon, 18 Sep 2006 08:50:14 +0200

electricsheep (2.6.8-2) unstable; urgency=low

  [ The "Hey babe, I'm a Debian Developer!" upload, part #1 ] 
  
  * Changed Maintainer field in debian/control to `sukria@debian.org'. 

 -- Alexis Sukrieh <sukria@debian.org>  Sat,  9 Sep 2006 10:27:36 +0200

electricsheep (2.6.8-1) unstable; urgency=low

  * New upstream release.
    (closes: 365652)

 -- Alexis Sukrieh <sukria@sukria.net>  Sat,  5 Aug 2006 10:43:00 +0200

electricsheep (2.6.6-1) unstable; urgency=low

  * New upstream release.
  * Fixed some packaging issues reported by Philipp Kern.
  * Updated autotools scripts:
    + Added AM_MAINTAINER_MODE to configure.in
    + Customized Makefile.am so we can drop debian/patches/01_Makefile.dpatch
  * debian/rules now uses DESTDIR instead of PREFIX, cleaner (thanks to
    Philipp Kern).
  * Policy revision: 3.7.2

 -- Alexis Sukrieh <sukria@sukria.net>  Tue,  2 May 2006 00:28:02 +0200

electricsheep (2.6.4-2) unstable; urgency=low

  * Added missing build-dependencies: libx11-dev, libxext-dev, x-dev, libxt-dev 
    (closes: #344653)

 -- Alexis Sukrieh <sukria@sukria.net>  Sun, 25 Dec 2005 10:26:31 +0100

electricsheep (2.6.4-1) unstable; urgency=low

  * New upstream release.
    + possible to preview electricsheep in xscreensaver while running other
    instances (closes: #273984).
    + possible to run electricsheep in standalone mode, with a specific
    datadir (closes: #341004).
  * Removed debian/electricsheep.postinst, not needed anymore with the new
    upstream code tree.

 -- Alexis Sukrieh <sukria@sukria.net>  Wed, 21 Dec 2005 15:49:57 +0100

electricsheep (2.6.3+cvs20051206-1) unstable; urgency=low

  * New upstream release (CVS snapshot from 2005/12/06)
    + Fixes several buffer overflows
  * New patch for canceling the use of xsetbg for avoiding bug #325689.
  * New configure.in for using AM_MAINTAINER_MODE.

 -- Alexis Sukrieh <sukria@sukria.net>  Tue,  6 Dec 2005 10:42:23 +0100

electricsheep (2.6.3-2) unstable; urgency=low

  * Removed the dependency to xlibs-dev, added libxv-dev instead which is the
    only one needed here.
  * Remove stubs maintainer scripts (they came from the old maintenance and I
    forgot to remove them, let's do a bit of cleaning).
    - debian/prerem
    - debian/postrm
  * Remove debian/REJECT (was containing the licence).
  * Remove the overkilling dependency xlibs-dev, the other build-dependencies
    are enough.
  * Changed the autotools scripts, using AM_MAINTAINER_MODE, added a
    bootstrap.
  * Removed duplicate debian/electricsheep.desktop file, is provided in
    debian/default-files.

 -- Alexis Sukrieh <sukria@sukria.net>  Tue, 29 Nov 2005 21:00:56 +0100

electricsheep (2.6.3-1) unstable; urgency=low

  * New upstream release.
    (closes: #318003)
  * New patch for Makefile.in.
  * Added a .desktop file for KDE (Thanks to Gonéri Le Bouder).
    (closes: #315420)

 -- Alexis Sukrieh <sukria@sukria.net>  Mon, 25 Jul 2005 14:12:57 +0200

electricsheep (2.6.2-1) unstable; urgency=low

  * New upstream release: 2.6.2.
    (closes: #309852)
  * Added myself as maintainer (Sander Smeenk offered it to me).
  * Added dpatch for handling changes applied on upstream sources.
  * Added all missing manpages.

 -- Alexis Sukrieh <sukria@sukria.net>  Fri, 17 Jun 2005 18:28:17 +0200

electricsheep (2.5-2) unstable; urgency=low

  + Fixed builddepends on libpng (Closes: #245816)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 25 Apr 2004 17:46:57 +0200

electricsheep (2.5-1) unstable; urgency=low

  + New Upstream Release (Closes: #243760)
  + vroot.h was updated which Closes: #217492
  
 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 18 Apr 2004 15:06:34 +0200

electricsheep (2.4-3) unstable; urgency=low
  
  + Fixed dephelper depends (Closes: #216014)
  + Fixed conflicts with libmpeg2-3-dev (Closes: #207342)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 19 Oct 2003 15:59:02 +0200

electricsheep (2.4-2) unstable; urgency=low

  + Correct dependencies (Closes: #206346)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 24 Aug 2003 12:31:00 +0200
   
electricsheep (2.4-1) unstable; urgency=low

  + New upstream version (Closes: #203965)

 -- Sander Smeenk <ssmeenk@debian.org>  Thu, 14 Aug 2003 10:18:40 +0200

electricsheep (2.3-2) unstable; urgency=low

  + Moved *tif out of /usr/share to /usr/share/electricsheep.
  + Forgot to close bugs with my previous upload.
    Closes: #192510
  + Buildproblems on noni386 are fixed upstream
    Thanks Julien Lemoine for your effort.
    Closes: #167963, #182510
  + Added build dependency to libexpat1-dev (Closes: #185390)
  + Actually using CFLAGS from debian/rules now

 -- Sander Smeenk <ssmeenk@debian.org>  Tue, 18 Mar 2003 14:44:08 +0100

electricsheep (2.3-1) unstable; urgency=low

  + New upstream.
  + New debian/ structure, new Standards version, etc.

 -- Sander Smeenk <ssmeenk@debian.org>  Wed, 26 Feb 2003 09:26:08 +0100

electricsheep (2.1-6) unstable; urgency=low

  + 'electricsheep.org' the domain that pointed to the official

electricsheep-voter.c
/*
    electricsheep - collaborative screensaver
    Copyright (C) 1999-2003 Scott Draves <source@electricsheep.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

static char *electricsheep_voter_c_id =
"@(#) $Id: electricsheep-voter.c,v 1.8 2005/07/11 07:22:17 spotspot Exp $";

/*
 * When an arrow key is pressed transmit a vote to the server.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/param.h>
#include "config.h"

char *client_version = "LNX_" VERSION;

/* instead of watching for X events, watch for
   griffin powermate button presses */
#define POWERMATE   0

/* instead of wathing for X events,
   just pop up logos */
#define FLYINGLOGOS 0


#if POWERMATE
#include <linux/input.h>
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#elif !FLYINGLOGOS
#include <X11/Xlib.h>
#include <X11/keysym.h>
#include "mpeg2dec/libvo/vroot.h"
#endif

char *nick, *url, *uid;
char *history_file;
char *dream_server;
char *leave_prefix;

char *splash_prefix = PACKAGE_DATA_DIR "/electricsheep";

void vote(int how) {
    FILE *hf;
    char sheep_name[1000];
    char vote_file_name[MAXPATHLEN];
    char pbuf[MAXPATHLEN];
    hf = fopen(history_file, "rb");
    if (NULL == hf) {
	perror(history_file);
	return;
    }
    if (NULL == fgets(sheep_name, 1000, hf)) {
	fprintf(stderr, "could not read sheep name from history file.\n");
	fclose(hf);
	return;
    }
    fclose(hf);
    if (strlen(sheep_name) > 1) {
	sheep_name[strlen(sheep_name)-1] = 0; /* chop newline */

	if (0 == fork()) {
	  sprintf(pbuf, "echo '!%s-%s.png 10 30 30' > "
		  "%soverlay_fifo", splash_prefix,
		  (how>0)?"smile":"frown", leave_prefix);
	  system(pbuf);
	  exit(0);
	}
	sprintf(pbuf, "curl 'http://%s/cgi/vote?id=%s&vote=%d&u=%s&v=%s' >& /dev/null",
		dream_server, sheep_name, how, uid, client_version);
	if (0 == fork()) {
	  system(pbuf);
	  exit(0);
	}
    } else {
	fprintf(stderr, "empty sheep name.\n");
    }
}

#if POWERMATE

#define NUM_VALID_PREFIXES 2

static const char *valid_prefix[NUM_VALID_PREFIXES] = {
  "Griffin PowerMate",
  "Griffin SoundKnob"
};

#define NUM_EVENT_DEVICES 16

int open_powermate(const char *dev, int mode)
{
  int fd = open(dev, mode);
  int i;
  char name[255];

  if(fd < 0){
    fprintf(stderr, "Unable to open \"%s\": %s\n", dev, strerror(errno));
    return -1;
  }

  if(ioctl(fd, EVIOCGNAME(sizeof(name)), name) < 0){
    fprintf(stderr, "\"%s\": EVIOCGNAME failed: %s\n", dev, strerror(errno));
    close(fd);
    return -1;
  }

  // it's the correct device if the prefix matches what we expect it to be:
  for(i=0; i<NUM_VALID_PREFIXES; i++)
    if(!strncasecmp(name, valid_prefix[i], strlen(valid_prefix[i])))
      return fd;

  close(fd);
  return -1;
}

int abs_offset = 0;

void process_event(struct input_event *ev)
{
#ifdef VERBOSE
  fprintf(stderr, "type=0x%04x, code=0x%04x, value=%d\n",
	  ev->type, ev->code, (int)ev->value);
#endif

  switch(ev->type){
  case EV_KEY:
    if (ev->value) vote(1);
    break;
  }
}

#define BUFFER_SIZE 32
void watch_powermate(int fd)
{
  struct input_event ibuffer[BUFFER_SIZE];
  int r, events, i;

  while(1){
    r = read(fd, ibuffer, sizeof(struct input_event) * BUFFER_SIZE);
    if( r > 0 ){
      events = r / sizeof(struct input_event);
      for(i=0; i<events; i++)
	process_event(&ibuffer[i]);
    }else{
      fprintf(stderr, "read() failed: %s\n", strerror(errno));
      return;
    }
  }
}
#endif

#if POWERMATE
int find_powermate(int mode)
{
  char devname[256];
  int i, r;

  for(i=0; i<NUM_EVENT_DEVICES; i++){
    sprintf(devname, "/dev/input/event%d", i);
    r = open_powermate(devname, mode);
    if(r >= 0)
      return r;
  }

  return -1;
}
#endif

void fly_logo(char *cmd) {
  char buf[MAXPATHLEN];
  sprintf(buf, "echo '%s' > %soverlay_fifo", cmd, leave_prefix);
  system(buf);
}
  

int main(int argc, char **argv) {
  int window;
#if POWERMATE
  int powermate = -1;
#elif !FLYINGLOGOS
  Display *display;
  XWindowAttributes xgwa;
#endif

  if (argc != 8) {
    printf("usage: electricsheep-voter window nick url "
	   "history_file dream_server leave_prefix uid\n");
    printf("%d\n", argc);
    return 0;
  }
  window = strtol(argv[1], NULL, 0);
  nick = argv[2];
  url = argv[3];
  history_file = argv[4];
  dream_server = argv[5];
  leave_prefix = argv[6];
  uid = argv[7];

#if POWERMATE
    

  powermate = find_powermate(O_RDONLY);


  if(powermate < 0){
    fprintf(stderr, "Unable to locate powermate\n");
    return 1;
  }

  watch_powermate(powermate);

  close(powermate);

#elif  FLYINGLOGOS
  while (1) {
    sleep(10);
    fly_logo("!/home/sheep/smile.ppm 30 60 30");
    sleep(10);
    fly_logo("!/home/sheep/frown.ppm 30 60 30");
  }
#else
  display = XOpenDisplay(NULL);
  if (NULL == display) {
    fprintf(stderr, "error: cannot open display.\n");
    return EXIT_FAILURE;
  }

  if (window == -3 || window == -2) {
    /* monitor the (virtual) root window */
    window = DefaultRootWindow (display);
  } else if (window == -1) {
    fprintf(stderr, "error: voting not allowed in separate window.\n");
    return EXIT_FAILURE;
  }

  XGetWindowAttributes(display, window, &xgwa);
  xgwa.your_event_mask = KeyPressMask;
  XSelectInput(display, window, xgwa.your_event_mask);

  while (1) {
    XEvent event;
    char s[2];
    int size;
    KeySym keysym;
    XNextEvent(display, &event);
    if (event.type != KeyPress) continue;
    size = XLookupString ((XKeyEvent *) &event, s, 1, &keysym, 0);
    switch (keysym) {
    case XK_Up:
      vote(1);
      break;
    case XK_Down:
      vote(-1);
      break;
    }
  }

#endif

  /* not reached */
  return EXIT_SUCCESS;
}
